packages/test/test-end-to-end-tests/src/test/dataMigration.spec.ts

```diff
@@ -0,0 +1,425 @@
+	) {
+		super();
+	}
+	public async preInitialize(
+		context: IContainerContext,
+		existing: boolean,
+	): Promise<ContainerRuntime> {
+		this.clientDetailsType = context.clientDetails.type;
+		const codeDetails = context.getSpecifiedCodeDetails
+			? context.getSpecifiedCodeDetails()
+			: undefined;
+		assert(codeDetails !== undefined, "get code details failed");
+
+		this._currentRuntimeFactory = codeDetails.package === "v1" ? this.v1 : this.v2;
+		if (this.clientDetailsType === detachedClientType) {
+			const detachedContext = createReadOnlyContext(context, "v2");
+			this._currentRuntimeFactory = this.v1ToV2;
+			return this.currentRuntimeFactory.preInitialize(detachedContext, existing);
+		}
+
+		return this.currentRuntimeFactory.preInitialize(context, existing);
+	}
+
+	public async instantiateFirstTime(runtime: ContainerRuntime): Promise<void> {
+		return this.currentRuntimeFactory.instantiateFirstTime(runtime);
+	}
+
+	public async instantiateFromExisting(runtime: ContainerRuntime): Promise<void> {
+		return this.currentRuntimeFactory.instantiateFromExisting(runtime);
+	}
+
+	public async hasInitialized(runtime: ContainerRuntime): Promise<void> {
+		await this.currentRuntimeFactory.hasInitialized(runtime);
+
+		if (this.clientDetailsType === summarizerClientType) {
+			this.oldRuntime = runtime;
+			const readonlyRuntime = await runtime.loadDetachedAndTransitionFn();
+			this.deferred.resolve({ summarizer: this.oldRuntime, readonlyRuntime });
+		}
+	}
+}
+
+class SharedCellToSharedMapFactory implements IChannelFactory {
+	public static readonly Type = "https://graph.microsoft.com/types/cell";
+
+	public get type(): string {
+		return SharedCellToSharedMapFactory.Type;
+	}
+
+	public get attributes(): IChannelAttributes {
+		return SharedMap.getFactory().attributes;
+	}
+
+	public async load(
+		runtime: IFluidDataStoreRuntime,
+		id: string,
+		services: IChannelServices,
+		attributes: IChannelAttributes,
+	): Promise<SharedMap> {
+		const cell = new SharedCell(id, runtime, attributes);
+		await cell.load(services);
+		const map = SharedMap.getFactory().create(runtime, id) as SharedMap;
+		const val = cell.get();
+		map.set(cellKey, val as string);
+		return map;
+	}
+
+	public create(document: IFluidDataStoreRuntime, id: string): SharedMap {
+		throw new Error("Shouldn't be making a shared cell");
+	}
+}
+
+async function logTree(tree: ISnapshotTree, storage: IDocumentStorageService) {
+	const newTree: any = {};
+	newTree.blobs = {};
+	newTree.trees = {};
+	for (const [blobName, blobId] of Object.entries(tree.blobs)) {
+		const content = await readAndParse(storage, blobId);
+		newTree.blobs[blobName] = content;
+	}
+	for (const [treeName, treeNode] of Object.entries(tree.trees)) {
+		newTree.trees[treeName] = await logTree(treeNode, storage);
+	}
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-return
+	return newTree;
+}
+
+async function migrate(summarizerRuntime: ContainerRuntime, readonlyRuntime: ContainerRuntime) {
```

Comment:
To be honest, I got lost here on what these runtimes are. Trying to follow the code, it feels like readonlyRuntime should be a detached runtime, but if so, I'm not sure why we call startSummary() on it, so I think I got it wrong (and thus got lost)

Reply:
So if we call startSummary, we can call summarize and potentially get an incremental summary that is submittable.

Reply:
If we don't call startSummary, we can't summarize because the summarizerNode system would break down. All clients utilize the summarizerNode system from what I understand. This may change in the future, I'm not sure.